### PR TITLE
Add Support for Retrieving PVT Property Curves

### PR DESCRIPTION
**Pull request by bska**
This change-set defines a simple protocol for retrieving the "raw" PVT property curves directly from an ECL INIT file.  At present we recognise requests for the formation volume factor and viscosity curves from immiscible tables (i.e., those represented by input keywords `PVDG` and `PVDO`).

Water properties are not supported at this time.

Note as well that the common requests generate instant runtime failure (`std::runtime_error` exception) if the backing function is from a miscible table.  This is, obviously, not a long-term solution.

Finally, we add a wrapper class `ECLPvtCurveCollection` that simplifies retrieving PVT curves for multiple phases in a single active cell.

### CODE REVIEW

File: opm/utility/ECLPvtCurveCollection.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        // Active cell index out of bounds.  Return empty.
+        return FlowDiagnostics::Graph{};
+    }
+
+    // PVTNUM is traditional one-based region identifier.  Subtract one to
+    // form valid index into std::vector<>s.
+    const auto regID = this->pvtnum_[activeCell] - 1;
+
+    if (phase == ECLPhaseIndex::Liquid) {
+        // return this->oil_->getPvtCurve(curve, regID);
```
Comment (atgeirr) [inline opm/utility/ECLPvtCurveCollection.cpp]: Did you forget to uncomment this?

Comment (bska) [reply opm/utility/ECLPvtCurveCollection.cpp]: > Did you forget to uncomment this?

Eh, yes.  Sorry about that.

**Comment by bska** [comment]
>```
> bin/extractPropCurves case=~/opm/opm-data/spe9/eclipse-simulation/SPE9_CP krog=true
>```
> yields a curve that I cannot recognize from the input deck data. (The krg table looks right though.)

What, precisely, do you mean by "not recognize"?  That there are more/other saturation points or that the shape/range is incorrect?

Do keep in mind that relative permeabilities for oil in the INIT files are returned as functions of increasing oil saturation--essentially as if entered using `SOF2` or `SOF3`.  Moreover, since SPE9_CP uses family I (`SWOF`  and `SGOF`), the two tables are merged and missing values calculated by piecewise linear interpolation.  I am definitely not claiming that this is fully bug-free, but I am saying that there is a little process that goes into creating the normalised saturation function tables in the INIT file.

**Comment by atgeirr** [comment]
This is the curve I had trouble recognizing:
```
krog = [
0.0000000000000000e+00 0.0000000000000000e+00
1.1851000000000000e-01 0.0000000000000000e+00
1.4891000000000010e-01 0.0000000000000000e+00
1.8889000000000000e-01 0.0000000000000000e+00
2.4891000000000008e-01 0.0000000000000000e+00
2.8720000000000001e-01 0.0000000000000000e+00
3.4306999999999999e-01 0.0000000000000000e+00
3.4891000000000005e-01 0.0000000000000000e+00
3.9893999999999996e-01 0.0000000000000000e+00
4.2688000000000004e-01 0.0000000000000000e+00
4.4891000000000003e-01 0.0000000000000000e+00
4.8275000000000001e-01 6.7679999999999980e-03
5.1068000000000002e-01 1.2354000000000000e-02
5.3861000000000003e-01 1.7940000000000005e-02
5.4891000000000001e-01 2.0000000000000000e-02
5.6655000000000000e-01 3.4111999999999976e-02
5.9999999999999998e-01 6.0871999999999940e-02
6.2000000000000000e-01 7.6871999999999940e-02
6.3000000000000000e-01 8.4871999999999947e-02
6.3121000000000005e-01 8.5839999999999986e-02
6.3560499999999998e-01 8.9355999999999922e-02
6.4000000000000001e-01 9.2871999999999941e-02
6.4200000000000002e-01 9.4471999999999945e-02
6.4400000000000002e-01 9.6071999999999949e-02
6.4600000000000002e-01 9.7671999999999953e-02
6.4800000000000002e-01 9.9271999999999944e-02
6.4891000000000010e-01 1.0000000000000001e-01
6.5000000000000002e-01 1.0250699999999983e-01
7.4665000000000004e-01 3.2480199999999992e-01
7.4891000000000008e-01 3.3000000000000002e-01
7.9664999999999997e-01 5.4482999999999970e-01
8.0891000000000002e-01 5.9999999999999998e-01
8.2164999999999999e-01 7.2739999999999960e-01
8.3414999999999995e-01 8.5239999999999905e-01
8.4353000000000000e-01 9.4619999999999949e-01
8.4753999999999996e-01 9.8629999999999907e-01
8.4826000000000001e-01 9.9349999999999961e-01
8.4877000000000002e-01 9.9859999999999971e-01
8.4891000000000005e-01 1.0000000000000000e+00
];
```
vs the third column from the deck:
```
SGOF
0  	  0	     1		0
0.04	  0	     0.6	0.2
0.1	  0.022	     0.33	0.5
0.2	  0.1	     0.1	1.0
0.3	  0.24	     0.02	1.5
0.4	  0.34	     0		2.0
0.5	  0.42	     0		2.5
0.6	  0.5	     0		3.0
0.7	  0.8125     0		3.5
0.84891   0.9635     0		3.82 /
```

**Comment by bska** [comment]
> This is the curve I had trouble recognizing [`krog` is 39-point table] vs the third column from the deck [`krog` is 10-point table].

Right.  However, the simulator merges the tabulated relative permeabilities for oil in both the oil-gas (`krog`) and oil-water (`krow`) systems into the **same** table as a preprocessing step, so there is going to be one saturation point for each unique oil saturation value in the union of (1-Sw) from `SWOF` and `1-Sg-Swco` from `SGOF`.  Since the `SWOF` table for SPE9_CP is so large (30 points), that is typically going to generate many points that don't have a corresponding oil saturation in `SGOF`.  When using familiy I, it is quite common for the size of the oil saturation functions to be essentially the sum of the sizes of the `SGOF` and `SWOF` tables.

In other words, you cannot normally compare the `SOFN` values from the INIT file (essentially `SOF3`) directly to the tables if the deck uses `SWOF` and `SGOF`, but at minimum we should preserve the shape of the tabulated functions and all unique saturation points in both tables should be present in the normalized output file.

**Comment by atgeirr** [comment]
I know that there would be extra values, but there is something I don't understand here: how can the simulator put both values into the same table? I assume that you mean as separate columns, as the SOF3 keyword?

With that assumption, the only discrepancy seems to be that the "krog" table is shifted by the connate water saturation. Should it be?

**Comment by bska** [comment]
> how can the simulator put both values into the same table? I assume that you mean as separate columns, as the `SOF3` keyword?

Yes.  As a preprocessing step, the simulator creates a table called `SOFN` that looks like
```
So   krow   krog
 :    :      :
```
in a three-phase run.  In a two-phase run, the `SOFN` table looks like
```
So   kro
 :    :
```

If the simulation run is specified "Family II" keywords (i.e., `SOF2` or `SOF3` and `SGFN` and/or `SWFN`) then the `SOFN` table is just a copy of the input oil table.  If, on the other hand, the simulation run uses "Family I" (`SWOF` and/or `SGOF`), the `SOFN` table results from merging the sorted `So` ranges, picking the tabulated "Kr" values from the appropriate table and replacing the missing "Kr" values (the ones in the "other" table) with quantities calculated by piecewise linear interpolation.

> With that assumption, the only discrepancy seems to be that the "krog" table is shifted by the connate water saturation. Should it be?

Yes.  The sorted `So` ranges are `1-Sw` from `SWOF` and `(1-Sg)-Swco` with `Sg` from `SGOF` and `Swco` the connate water saturation.
